internal/app/machined/pkg/system/services/apid.go

```diff
@@ -0,0 +1,155 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// nolint: dupl,golint
+package services
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	containerdapi "github.com/containerd/containerd"
+	"github.com/containerd/containerd/oci"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+
+	"github.com/talos-systems/talos/internal/app/machined/pkg/system/conditions"
+	"github.com/talos-systems/talos/internal/app/machined/pkg/system/health"
+	"github.com/talos-systems/talos/internal/app/machined/pkg/system/runner"
+	"github.com/talos-systems/talos/internal/app/machined/pkg/system/runner/containerd"
+	"github.com/talos-systems/talos/internal/app/machined/pkg/system/runner/restart"
+	"github.com/talos-systems/talos/internal/pkg/runtime"
+	"github.com/talos-systems/talos/pkg/config/machine"
+	"github.com/talos-systems/talos/pkg/constants"
+	"github.com/talos-systems/talos/pkg/kubernetes"
+	"github.com/talos-systems/talos/pkg/retry"
+)
+
+// APID implements the Service interface. It serves as the concrete type with
+// the required methods.
+type APID struct{}
+
+// ID implements the Service interface.
+func (o *APID) ID(config runtime.Configurator) string {
+	return "apid"
+}
+
+// PreFunc implements the Service interface.
+func (o *APID) PreFunc(ctx context.Context, config runtime.Configurator) error {
+	importer := containerd.NewImporter(constants.SystemContainerdNamespace, containerd.WithContainerdAddress(constants.SystemContainerdAddress))
+
+	return importer.Import(&containerd.ImportRequest{
+		Path: "/usr/images/apid.tar",
+		Options: []containerdapi.ImportOpt{
+			containerdapi.WithIndexName("talos/apid"),
+		},
+	})
+}
+
+// PostFunc implements the Service interface.
+func (o *APID) PostFunc(config runtime.Configurator) (err error) {
+	return nil
+}
+
+// Condition implements the Service interface.
+func (o *APID) Condition(config runtime.Configurator) conditions.Condition {
+	if config.Machine().Type() == machine.Worker {
+		return conditions.WaitForFileToExist(constants.KubeletKubeconfig)
+	}
+
+	return nil
+}
+
+// DependsOn implements the Service interface.
+func (o *APID) DependsOn(config runtime.Configurator) []string {
+	if config.Machine().Type() == machine.Worker {
+		return []string{"system-containerd", "containerd", "kubelet"}
```

Comment:
I don't think we need the dependency on `kubelet`.

Reply:
osd depended on kubelet to make use of admin.conf to query for master nodes. i suppose we could drop the dependency for apid and not have os api available until osd comes up. i'll drop it.